src/core/mount.c

```diff
@@ -1280,23 +1280,20 @@ static void mount_sigchld_event(Unit *u, pid_t pid, int code, int status) {
         log_unit_full(u, f == MOUNT_SUCCESS ? LOG_DEBUG : LOG_NOTICE, 0,
                       "Mount process exited, code=%s status=%i", sigchld_code_to_string(code), status);
 
-        /* Note that mount(8) returning and the kernel sending us a mount table change event might happen
-         * out-of-order. If an operation succeed we assume the kernel will follow soon too and already change into the
-         * resulting state.  If it fails we check if the kernel still knows about the mount. and change state
-         * accordingly. */
+        /* Note that due to the io event priority logic, we can be sure the new mountinfo is loaded
+         * before we process the SIGCHLD for the mount command. */
 
         switch (m->state) {
 
         case MOUNT_MOUNTING:
-        case MOUNT_MOUNTING_DONE:
+                /* No new mountinfo for this mount point.
+                 * Someone raced us and we lost (or /bin/mount lies to us and is broken). */
+                log_unit_warning(UNIT(m), "Mount process finished, but there is no mount.");
+                mount_enter_dead(m, f);
```

Comment:
hmm, i figure we should define a new error result here for the case where /bin/mount exited cleanly but where it didn't actually establish any mount point. i.e. the MountResult enum should gain MOUNT_FAILURE_PROTOCOL or so, and then we should change the line above to:

```c
mount_enter_dead(m, f == MOUNT_SUCCESS ? MOUNT_FAILURE_PROTOCOL : f);
```

after all, /bin/mount broke protocol here, and we should report that

Reply:
specifically, it makes sure mount_enter_dead() sets the unit to a failed state.  Excellent point, that definitely seems more correct.